Allow `&&`, `||`, and `!` in `cfg`

jhpratt opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rfcs/pull/3796)*

About as simple as it gets.

[Rendered](https://github.com/jhpratt/rfcs/blob/cfg-logical-ops/text/0000-cfg-logical-ops.md)

jhpratt committed:
Start cfg-logical-ops

programmerjake commented on text/0000-cfg-logical-ops.md:
```diff
@@ -0,0 +1,94 @@
+> &nbsp;&nbsp; (`not` | `!`) `(` _ConfigurationPredicate_ `)`
+>
+> _ConfigurationAnd_\
+> &nbsp;&nbsp; _ConfigurationPredicateList_ `&&` _ConfigurationPredicateList_
+>
+> _ConfigurationOr_\
+> &nbsp;&nbsp; _ConfigurationPredicateList_ `||` _ConfigurationPredicateList_
```
I expect `a, b && c, d` to not be valid where `a, b` is not valid, also I expect `a, b && c, d` to not mean `(a, b) && (c, d)`

```suggestion
> &nbsp;&nbsp; _ConfigurationPredicate_ `&&` _ConfigurationPredicate_
>
> _ConfigurationOr_\
> &nbsp;&nbsp; _ConfigurationPredicate_ `||` _ConfigurationPredicate_
```

jhpratt replied:
funnily enough I had this originally but misread it when copying into the final format

bluebear94 commented on text/0000-cfg-logical-ops.md:
```diff
@@ -0,0 +1,94 @@
+> &nbsp;&nbsp; _ConfigurationPredicateList_ `||` _ConfigurationPredicateList_
+>
+> _ConfigurationNotOption_\
+> &nbsp;&nbsp; `!` _ConfigurationOption_
```
This would allow things like `#[cfg(!feature = "foo")]`, which looks strange. Could we restrict this to taking options that consist of a single identifier?

CraftSpider replied:
At least personally, I use `key = "value"` predicates fairly frequently, so I'd expect them to work. I haven't looked closely at the grammar, but if the choice is made to not allow `!feature = "foo"`, I'd hope to at least keep `!(feature = "foo")` or similar, so I can actually use the `!` syntax basically ever.

jhpratt replied:
Yes, this would permit that, which was a deliberate decision. If there is not the will for it, I can remove it or restrict per your suggestion, leaving only `!(feature = "foo")`. That would be permitted by the `ConfigurationNot` rule.

RalfJung replied:
I feel fairly strongly that we should not allow `!feature = "foo"`. This part of the syntax goes against existing precedent in Rust and is visually quite confusing. In code we write `!(a == b)`, or `a != b`, so that's what we should do here as well.

jhpratt replied:
I'll remove this from the text later today, requiring parentheses in this situation.

jhpratt committed:
Fix mistake in and/or syntax

kennytm commented on text/0000-cfg-logical-ops.md:
```diff
@@ -0,0 +1,94 @@
+1.34, but the `cfg` syntax was not updated to take advantage of this. By letting developers use
+logical operators, we are _lessening_ the burden of having to remember the `cfg` syntax.
+
+# Explanation
```
i'd like to see some more complex examples demonstrating what is supported and invalid, such as the strange `#[cfg(!feature = "foo")]`

| Syntax | Expected lowering | Rationale |
|---|---|---|
| `a && b` | `all(a, b)` | Definition of `&&` |
| `a \|\| b` | `any(a, b)` | Definition of `\|\|` |
| `!a` | `not(a)` | Definition of `!` |
| `(a)` | `a` | Definition of `()` |
| `a && b && c && d` | `all(a, b, c, d)` (or `all(all(all(a, b), c), d)`) | `&&` is associative |
| `a \|\| b \|\| c \|\| d` | `any(a, b, c, d)` (or `any(any(any(a, b), c), d)`) | `\|\|` is associative |
| `!!!!!!a` | `not(not(not(not(not(not(a))))))` | `!` can be nested |
| `((((((a))))))` | a | `()` can be nested |
| `a && b \|\| c && d` | `any(all(a, b), all(c, d))` | `\|\|` has lower precedence than `&&` |
| `a \|\| b && c \|\| d` | `any(a, all(b, c), d)` | `\|\|` has lower precedence than `&&` |
| `(a \|\| b) && (c \|\| d)` | `all(any(a, b), any(c, d))` | `()` can be used for grouping |
| `!a \|\| !b && !c` | `any(not(a), all(not(b), not(c)))` | `!` has highest precedence |
| `feature="foo" \|\| feature="bar"` | `any(feature="foo", feature="bar")` | `\|\|` has lower precedence than `=` |
| `feature="foo" && feature="bar"` | `all(feature="foo", feature="bar")` | `&&` has lower precedence than `=` |
| `!feature="foo"` | syntax error | `!` has higher precedence than `=`, which is confusing, so we ban this syntax |
| `!(feature="foo")` | `not(feature="foo")` | Use `()` for grouping |
| `!all(x, y)` | `not(all(x, y))` | `!` has lower precedence than "function call" |
| `any(!x \|\| !w, !(y && z))` | `any(any(not(x), not(w)), not(all(y, z)))` | `!`, `&&` etc can be used inside `any`, `all` and `not` |
| `true && !false` | `all(true, not(false))` | `!`, `&&` etc can be used on boolean literals |
| `!accessible(std::mem::forget)` | `not(accessible(std::mem::forget))` | `!`, `&&` etc can be used on cfg_accessible |
| `accessible(std::a \|\| std::b)` | syntax error | ... but not inside |
| `!version("1.42.0")` | `not(version("1.42.0"))` | `!`, `&&` etc can be used on cfg_version |
| `version(!"1.42.0")` | syntax error | ... but not inside |

jhpratt commented on text/0000-cfg-logical-ops.md:
```diff
@@ -0,0 +1,94 @@
+> _ConfigurationNot_\
+> &nbsp;&nbsp; (`not` | `!`) `(` _ConfigurationPredicate_ `)`
+>
+> _ConfigurationAnd_\
+> &nbsp;&nbsp; _ConfigurationPredicate_ `&&` _ConfigurationPredicate_
+>
+> _ConfigurationOr_\
+> &nbsp;&nbsp; _ConfigurationPredicate_ `||` _ConfigurationPredicate_
```
I was almost positive this was required coming in. Is there a "typical" way of expressing this in formal grammar? Easier to ask than spend time figuring it out myself.

CosmicHorrorDev replied:
I believe you would define the grammar to follow transitions through lower precedence to higher, so something like. An example with just `||`, `&&`, `!`, and `()` (in the order of increasing precedence)

With the starting state of ConfigurationPredicate

_Transitions to climb precedence with a single value alone:_
- ConfigurationPredicate -> ConfigurationOr
- ConfigurationOr -> ConfigurationAnd
- ConfigurationAnd -> ConfigurationNot
- ConfigurationNot -> ConfigurationParens

_And the actual transitions with operators:_
- ConfigurationOr -> ConfigurationOr `||` ConfigurationAnd
- ConfigurationAnd -> ConfigurationAnd `&&` ConfigurationNot
- ConfigurationNot -> `!` ConfigurationParens
- ConfigurationParens -> `(` ConfigurationOr `)`

_Terminating the grammar with idents:_
- ConfigurationParens -> Ident

---

Consider `a && b || c && (d || !e)`

1. ConfigurationPredicate -> ConfigurationOr
2. ConfigurationOr -> ConfigurationOr `||` ConfigurationAnd with `a && b` `||` `c && (d || !e)`

Considering the chunk at ConfigurationOr with `a && b`

1. ConfigurationOr -> ConfigurationAnd
2. ConfigurationAnd -> ConfigurationAnd `&&` ConfigurationNot with `a` `&&` `b`
3. `a` and `b` go through all of the transitions to Ident

Considering the chunk at ConfigurationAnd with `c && (c || !e)`

1. ConfigurationAnd -> ConfigurationAnd `&&` ConfigurationNot with `c` `&&` `(c || !e)`
2. `c` goes all the way to Ident

Considering the chunk at ConfigurationNot with `(c || !e)`

1. ConfigurationNot -> ConfigurationParens
2. ConfigurationParens -> `(` ConfigurationOr `)` with `(` `c || !e` `)`
3. ConfigurationOr -> ConfigurationOr `||` ConfigurationAnd with `c` `||` `!e`
4. ConfigurationAnd -> ConfigurationNot
5. ConfigurationNot -> `!` ConfigurationParens with `!` `e`
6. ConfigurationParens -> Ident with `e`

This grammar avoids ambiguity (in the formal sense of only having a single tree that can represent a given expression) and handles precedence correctly (mentally you can throw parens around the chunks to see that they get grouped right. It'd be easier to show as a tree, but that's a lot of effort)

programmerjake replied:
rust's expression grammar in the rust reference doesn't use that method to avoid ambiguity and handle precedence, instead it just has a table for expression precedence: https://doc.rust-lang.org/reference/expressions.html#expression-precedence

CosmicHorrorDev replied:
A table definitely sounds a lot simpler :grin: 

Thanks, I'm not familiar with the project specifics, so that seems like a better approach

jhpratt replied:
I can explicitly state that `&&` has higher precedence, leaving the grammar as-is from the sounds of it?

programmerjake replied:
maybe just say it has the same precedence rules as rust expressions? if you want an explicit table, maybe:
| Operator/Punctuation | Associativity |
|-|-|
| parenthesis and `all(...)`/`any(...)`/`not(...)` | &nbsp; |
| `!` operator | &nbsp; |
| `&&` operator | left to right |
| `\|\|` operator | left to right |
| `=` syntax | &nbsp; |
| `,` punctuation | &nbsp; |

jhpratt replied:
writing things up now, and I'm simply linking to the reference

Jules-Bertholet commented on text/0000-cfg-logical-ops.md:
```diff
@@ -0,0 +1,94 @@
+# Summary
+[summary]: #summary
+
+`#[cfg]`, `#[cfg_attr]`, and `cfg!()` can use `&&`, `||`, and `!` for `all`, `any`, and `not`,
```
One alternative would be to use `&` and `|` instead of `&&` and `||`.

clarfonthey commented:
I think that a `feature != "string"` syntax as an alternative to `not(feature = "string")` would be a nice addition here. `!(feature = "string")` feels weirder to me than `not(feature = "string")`, honestly.

jhpratt commented:
The problem I see with `feature != "foo"` is that it could plausibly be interpreted as "any feature other than `foo`".

clarfonthey commented:
That interpretation feels like a stretch to me, but it's also understandable. Perhaps a `feature("string")` syntax might be a bit more natural with `!feature("string")`, and also extensible to `feature("string", "other string")`.

jhpratt committed:
update from feedback

jhpratt commented:
Updated the following:

- `!feature = "foo"` is no longer permitted. Negation can be followed by parentheses, a single identifier (so `#[cfg(!unix)]`), and any function-like predicate.
- `&`/`|` added as alternative to `&&`/`||`
- `feature != "foo"` added as alternative to `!(feature = "foo")`
- added table with examples

programmerjake commented on text/0000-cfg-logical-ops.md:
```diff
@@ -20,13 +20,46 @@ logical operators, we are _lessening_ the burden of having to remember the `cfg`
-In terms of formal syntax, the [`[cfg.syntax]`][cfg-syntax] is changed to the following:
+enabled. `#[cfg_attr]` and `cfg!()` behave the same way.
+
+Precedence is the [same as in expressions][precedence].
```
you'll want to note `=` is treated as `==` for precedence purposes, since assignment has much lower precedence.

jhpratt replied:
done

jhpratt committed:
clarify `=` precedence

m-ou-se commented:
I read `#[cfg(feature = "asdf")]` similar to how I read things like `#[serde(default = "asdf")]` or `#[link_name = "asdf"]: The `=` syntax is used as a key-value pair, like in all attributes.

As mentioned above, `feature = "asdf"` is not at all like a Rust expression. It doesn't behave like a `=` expression (it's not an assignment). It doesn't behave like an `==` expression (it doesn't check equality either). There is no variable/concept named `feature` that is compared to anything.

It isn't even commutative: `#[cfg("asdf" = feature)]` doesn't work.

Using `key = "value"` syntax makes sense as it matches the syntax for all other attributes. But I don't think it makes much sense to take `&&` and `||` from Rust expressions and trying to force them into attribute syntax. The result feels weirdly inconsistent, with things like `!(feature = "asdf")` not being the same as `feature != "asdf"`, and so on.

So, if we want to go in this direction, we should also change the `feature = "asdf"` syntax to something that fits a Rust expression. E.g. `#[cfg(has_feature("asdf") && has_target_feature("asdf"))]`.

m-ou-se commented:
Let's look at a real world example:
```rust
#[cfg(any(target_os = "android", target_os = "linux", target_os = "emscripten", target_os = "freebsd"))]
```

This proposal would allow rewriting that as:

```rust
#[cfg(target_os = "android" || target_os = "linux" || target_os = "emscripten" || target_os = "freebsd"))]
```

Which looks a bit nicer.

But instead of just replacing `any()` by a series of `||`, I think we can do much better without giving up on the `key = value` syntax with something like:

```rust
#[cfg(target_os = "android" | "linux" | "emscripten" | "freebsd")]
```

I don't want to derail this thread with alternative proposals, but I do think that `&&` and `||` is not getting us that much. If we go in that direction of making cfg() accept expression-like syntax, then we might be closing ourselves off from improvements to the `key = value` syntax that might serve the actual use cases better.

RalfJung commented:
>  The result feels weirdly inconsistent, with things like !(feature = "asdf") not being the same as feature != "asdf"

How would they not be the same? The RFC, I think, intents for both of these to mean the same thing: the `feature` variable does not contain "asdf".

m-ou-se commented:
`!(feature = "asdf")` would work, while `feature != "asdf"` would be an error.

RalfJung commented:
Ah. That seems entirely consistent to me, it's `=` and not `==` after all.

I like `target_os = "android" | "linux" | "emscripten" | "freebsd"` but it doesn't when you need to look at more than one key.

m-ou-se commented:
My point is that we need tot look at real world examples of code we're trying to simplify.

For example, quite often, `any()` is used with the same key, such as when selecting for a list of different operating systems. We should optimize the syntax for patterns that occur in actual code.

Large cfg() attributes can be very confusing. But replacing `any()` and `all()` by `||` and `&&` isn't making it much better. Large boolean expressions will still be confusing, especially if their syntax is a hybrid of expressions and (key=value) attribute syntax with inconsistent precedence and operator meaning.

m-ou-se commented:
Rust parses `feature = "asdf" && unix` as `feature = ("asdf" && unix`)`, _not_ as `(feature = "asdf") && unix`.

One might argue that `=` in attributes is different, but how would we expect an attribute proc macro to parse this today?

```
#[my_attribute_macro(expr = a && b, expr2 = c || d)]
```

I would assume it interprets that as two `key = value` pairs with the values being `a && b` and `c || d`. (It's up to the proc macro to decide, this is what you get if you parse it with a Rust expression parser.)

So even in an attribute, I'd argue that `&&` should have higher precedence than the `key=value` operator.

programmerjake commented:
maybe embrace that they are sets and allow using `"abcd" in feature` instead of `feature = "abcd"`? though then that makes me think we should have named it `features`.

kennytm commented:
> I would go so far as to argue that `any()` and `all()` are _better for formatting_. Binary operators used repeatedly have an awkward indentation problem. 
>
> ...
>
> Whereas a similar boolean expression is less regular and longer:
>
> ...
>
> (And there are probably even messier cases to present with `&&` and `!` but I don't feel up to producing an illustrative example right now.) I won’t say that I haven’t wished for conventional binary operators in _simple_ cases, but switching to them completely would have significant disadvantages (as would having a mix of both as standard style, simply due to having two things to learn and the choice of which one to use).

Yes the formatting is strange, but I'd argue that the binary operator makes the `cfg` more *readable* for reviewers because you can be sure it is a chain of `any` rather than `all` at the first glance (spot the difference:

<table><tr><th>Existing syntax</th><td>

```rust
#[cfg(any(
    feature = "foo",
    feature = "bar",
    feature = "baz",
    feature = "quux",
    feature = "magic",
    test
))]
```

</td><td>

```rust
#[cfg(all(
    feature = "foo",
    feature = "bar",
    feature = "baz",
    feature = "quux",
    feature = "magic",
    test
))]
```

</td></tr><tr><th>This RFC</th><td>

```rust
#[cfg(
    (feature = "foo")
        || (feature = "bar")
        || (feature = "baz")
        || (feature = "quux")
        || (feature = "magic")
        || test
)]
```

</td><td>

```rust
#[cfg(
    (feature = "foo")
        && (feature = "bar")
        && (feature = "baz")
        && (feature = "quux")
        && (feature = "magic")
        && test
)]
```

</td></tr><tr><th>Alternative proposal</th><td>

```rust
#[cfg(any(
    feature = "foo" 
        | "bar" 
        | "baz" 
        | "quux" 
        | "magic",
    test,
))]
```
</td><td>

```rust
#[cfg(all(
    feature = "foo" 
        & "bar" 
        & "baz" 
        & "quux" 
        & "magic",
    test,
))]
```

</td></tr></table>

SabrinaJewson commented:
`cfg(any())` and `cfg(all())` are very useful, often in macros, as always-false and always-true `cfg` values respectively. This RFC does not _need_ to supplant all existing use cases of `all` and `any`, but it does seem weird to replace most things except this one edge case.

Jules-Bertholet commented:
> `cfg(any())` and `cfg(all())` are very useful, often in macros, as always-false and always-true `cfg` values respectively.

https://github.com/rust-lang/rust/issues/131204

joshtriplett commented:
@m-ou-se wrote:
> If we do want to make cfg() take something that looks like an expression, we should also change the `feature = "asdf"` syntax to something that fits a Rust expression. E.g. `#[cfg(has_feature("asdf") && has_target_feature("asdf"))]`.

Agreed completely. I don't think it would be ambiguous here to reuse "feature" without adding "has_": `cfg(feature("asdf"))`. And this would also be consistent with other parts of cfg syntax.

kennytm commented:
> 

> I don't think it would be ambiguous here to reuse "feature" without adding "has_": `cfg(feature("asdf"))`. And this would also be consistent with other parts of cfg syntax.

Technically this change of syntax will conflict with the cfg_version feature but since it is unstable it is fine...? (plus can't use `version="x"` on stable at all, it somehow triggers cfg_version too)

```rust
#![feature(cfg_version)]
fn main() {
	dbg!(cfg!(version("1.69.0")));
	dbg!(cfg!(version("2.0.0")));
	dbg!(cfg!(version = "1.69.0"));
	dbg!(cfg!(version = "2.0.0"));
}
```
```console
$ rustc +nightly --cfg 'version="2.0.0"' 1.rs
$ ./1
[1.rs:3:2] cfg!(version("1.69.0")) = true
[1.rs:4:2] cfg!(version("2.0.0")) = false
[1.rs:5:2] cfg!(version = "1.69.0") = false
[1.rs:6:2] cfg!(version = "2.0.0") = true
```

jdonszelmann commented:
I feel like many problems raised (by various people) here can be partially mitigated when we were to replace `feature = "something"` with `feature("something")` and equivalently for the other key-values such as `target_arch("x86_64")`  
I didn't come up with this, it felt like a slightly nicer way of writing @m-ou-se's `has_feature` which I like less, and now after having written the rest of this comment and reading back the thread I see that @clarfonthey proposed this quite literally.

Having said that, let this comment be a motivation for *why* I think this mitigates so many of the concerns. The advantages I see are:

1. The confusion about the meaning of  `=` (whether it means `==` or "assignment `=`" or a third kind of `=`) becomes irrelevant.
2. Because we use parentheses, we automatically disambiguate any use of `&&` and `||`. In other words, `#[my_attribute_macro(expr = a && b, expr2 = c || d)]` is instead written as  `#[my_attribute_macro(expr(a) && b && expr2(c) || d)]` Though of course, whether `&&` has precedence over `||` is still here
3. Whether to permit `feature != "a"` is irrelevant. it'd always be `!feature("a")`
4. it's consistent with `any` and `all` which are already parenthesized.
5. If we choose to permit alternation within an attribute (i.e. `feature = "a" | "b" | "c"` which with my proposal becomes `feature("a" | "b" | "c")`) this becomes clearly unambiguous from `feature("a") | b | c`

jhpratt commented:
I'll defer to the team here. I'm happy to write up whatever changes are desired, which at this point sounds like it may be `feature("foo")`, `key(val1 | val2 | val3)`, etc. If this is the case, I have one question: should we permit mixing and matching or only support all-or-nothing?

tmccombs commented:
I like the proposal in https://github.com/rust-lang/rfcs/pull/3796#issuecomment-2775430532, but the `feature = "abc"` syntax still needs to be supported for backwords compatibility, so we need to define how that interacts with the new syntax. Do we

1. Disallow the = syntax  in combination with && or ||
2. Require any use of = to be within parentheses, ex. `(feature="abc") && (target_os="linux")`
3. As originally described in this RFC
4. Something else

camsteffen commented:
I propose that we add `key(val | val)` syntax _without_ adding `&&`, `||`, `!`. That's where the bigger win is, and it would be less churn.

In a world of boolean logic, the existence of any/all/not functions is indubitably justified, and they deserve our respect.

jhpratt commented:
@camsteffen Is churn the primary (or even sole) concern? I think the :+1:s on this PR show that there's unquestionably the desire for logical operators — it feels obvious imo.

jdonszelmann commented:
I actually like @camsteffen's proposal a lot. It makes many scenarios simpler right now without changing ther original feature = name syntax. And the existing situation isn't so bad. And, if we feel like adding boolean operators later that still works!

camsteffen commented:
@jhpratt Not just churn, but the fact of having two ways to express something and a split in the ecosystem of which one is used. The motivation may be to make things simpler by matching other syntax, but I think the reality is that it's increasing complexity with another thing to learn. I would be more open to it if I could snap my fingers and change all the code in the world to the new syntax. But I also agree with @kpreid that any/all is better with regard to formatting.

Jules-Bertholet commented:
> > I would go so far as to argue that `any()` and `all()` are _better for formatting_. Binary operators used repeatedly have an awkward indentation problem…
> 
> Yes the formatting is strange, but I'd argue that the binary operator makes the `cfg` more _readable_ for reviewers…
>
> […]
>
> #### Alternative proposal	
> 
> ```rust
> #[cfg(any(
>     feature = "foo" 
>         | "bar" 
>         | "baz" 
>         | "quux" 
>         | "magic",
>     test,
> ))]
> ```
>
> ```rust
> #[cfg(all(
>     feature = "foo" 
>         & "bar" 
>         & "baz" 
>         & "quux" 
>         & "magic",
>     test,
> ))]
> ```


A little more indentation makes this look even better:

```rust
#[cfg(any(
    feature = "foo" 
            | "bar" 
            | "baz" 
            | "quux"
            | "magic",
    test,
))]
```

```rust
#[cfg(all(
    feature = "foo" 
            & "bar" 
            & "baz" 
            & "quux" 
            & "magic",
    test,
))]
```

kennytm commented:
@Jules-Bertholet the default rustfmt will not use vertical alignment.

----

@camsteffen Rust isn't afraid of having multiple ways to express the same concept, we have `for x in it {}` vs `while let Some(x) = t.next() {}`; `async fn f() -> R` vs `fn f() -> impl Future<Output=R>`; `fn g(a: impl Read)` vs `fn g<R: Read>(a: R)` vs `fn g<R>(a: R) where R: Read+,` etc. Even [your proposal](https://github.com/rust-lang/rfcs/pull/3796#issuecomment-2781102724) it going to introduce `#[cfg(feature = "name")]` vs `#[cfg(feature("name"))]`. IMO "splitting the ecosystem" is FUD and is no ground to reject a proposal for Rust.

This proposed syntax `#[cfg(key("val1" | "val2" | "val3"))]` works well when everything involved has the same "key", but won't help when multiple "keys" are involved, e.g.:

<table><thead><tr><td></td><th>

[Exhibit A](https://github.com/rust-lang/rust/blob/e643f59f6da3a84f43e75dea99afaa5b041ea6bf/library/std/src/os/mod.rs#L19-L25)

</td><th>

[Exhibit B](https://github.com/rust-lang/rust/blob/e643f59f6da3a84f43e75dea99afaa5b041ea6bf/library/unwind/src/lib.rs#L28-L34)

</td></tr></thead><tbody><tr><th>

Existing code

</th><td>

```rust
#[cfg(all(
    doc,
    any(
        all(target_arch = "wasm32", not(target_os = "wasi")),
        all(target_vendor = "fortanix", target_env = "sgx")
    )
))]
``` 

</td><td>

```rust
#[cfg(any(
    unix,
    windows,
    target_os = "psp",
    target_os = "solid_asp3",
    all(target_vendor = "fortanix", target_env = "sgx"),
))]
```

</td></tr><tr><th>This RFC</th><td>

```rust
#[cfg(doc && (
    (target_arch = "wasm32") && !(target_os = "wasi") 
        || (target_vendor = "fortanix") && (target_env = "sgx")
))]
```

</td><td>


```rust
#[cfg(unix 
    || windows 
    || (target_os = "psp") 
    || (target_os = "solid_asp3")
    || (target_vendor = "fortanix") && (target_env = "sgx")
)]
```

</td></tr><tr><th>key("value") only</th><td>

```rust
#[cfg(all(
    doc,
    any(
        all(target_arch("wasm32"), not(target_os("wasi"))),
        all(target_vendor("fortanix"), target_env("sgx"))
    )
))]
``` 

</td><td>

```rust
#[cfg(any(
    unix,
    windows,
    target_os("psp" | "solid_asp3"),
    all(target_vendor("fortanix"), target_env("sgx")),
))]
```

</td></tr><tr><th>This RFC + key("value")</th><td>

```rust
#[cfg(doc && (
    target_arch("wasm32") && !target_os("wasi") 
        || target_vendor("fortanix") && target_env("sgx")
))]
```

</td><td>

```rust
#[cfg(unix 
    || windows 
    || target_os("psp" | "solid_asp3")
    || target_vendor("fortanix") && target_env("sgx")
)]
```

</td></tr></tbody></table>

RalfJung commented:
Oddly, my first reaction to those examples that I'd like to use "any" and "&&"... I can't really explain why. But I do feel like the use of infix operators makes it less clear that we are looking at a list of cases.

RalfJung commented:
For those that are curious about the "nearly" part: the large family of ML-based languages including OCaml use `!` to mean "load from a reference" (that's a bit like `*` in Rust but not quite). "not" is written `~` in those languages (I think this is intended to be close the `¬`, the typical logical symbol for negation).

kornelski commented:
I really like this proposal. Not because it shortens some expressions, but because it's a step towards reducing the amount of unique microsyntaxes inside the Rust language.

kornelski commented:
As for the `=` bikeshed, how about `feature.name`/`target_os.wasi`, pretending that it's a struct with fields for all the options.

programmerjake commented:
> As for the `=` bikeshed, how about `feature.name`/`target_os.wasi`, pretending that it's a struct with `bool` fields for all the options.

what happens when values aren't a simple number/identifier? e.g. `--cfg=foo="abc def"`: https://rust.godbolt.org/z/bKfKGorbW

kennytm commented:
> As for the `=` bikeshed, how about `feature.name`/`target_os.wasi`, pretending that it's a struct with `bool` fields for all the options.

A [feature name](https://doc.rust-lang.org/cargo/reference/features.html#the-features-section) may consist of `[_0-9\p{XID_Start}][-+.\p{XID_Continue}]*`, and even the crates.io-restricted set is still `[_0-9a-zA-Z][-+_0-9a-zA-Z]*`. It would be very confusing for the parser to allow [`#[cfg(feature.c++20)]`](https://docs.rs/crate/cxx/latest/features).

Furthermore a custom config `--cfg 'thing="!@#$%^&"'` allows any string literal as the value (including empty string `--cfg 'thing=""'`, which is distinct from `--cfg 'thing'`). I don't think the quotation marks should be omitted.

```rust
#[cfg(feature = "c++20")]
#[cfg(feature("c++20")]
#[cfg(feature."c++20")]
```

programmerjake commented:
> I don't think the quotation marks should be omitted.

we could do something like TOML paths where you can omit the quotation marks if it's a valid identifier or number (equivalent to the corresponding quoted version) or you can have a string literal with any random text at all (except for commas since I think it's split into separate values using commas...).

so, with `--cfg=a="123" --cfg=b="abc" --cfg=c="123-4,abc"` that works with all of:
* `cfg(a.123)`
* `cfg(a."123")`
* `cfg(b.abc)`
* `cfg(b."abc")`
* `cfg(c."123-4")`
* `cfg(c.abc)`
* `cfg(c."abc")`

as well as all the existing `cfg(a = "123")` syntax.

kornelski commented:
Without inventing *new* syntax that isn't in Rust, this could also be `feature["#$%@!"]`.

RalfJung commented:
`key("value")` is not inventing new syntax.

Urgau commented on text/0000-cfg-logical-ops.md:
```diff
@@ -0,0 +1,146 @@
+| `!accessible(std::mem::forget)`    | `not(accessible(std::mem::forget))`                | `!`, `&&` etc. can be used on `cfg_accessible`                                      |
+| `accessible(std::a \|\| std::b)`   | _syntax error_                                     | … but not inside                                                                    |
+| `!version("1.42.0")`               | `not(version("1.42.0"))`                           | `!`, `&&` etc. can be used on `cfg_version`                                         |
+| `version(!"1.42.0")`               | _syntax error_                                     | … but not inside                                                                    |
```
How would it interact with the `target(os = "linux", ...)` shorthand from [RFC 3239](https://github.com/rust-lang/rfcs/pull/3239)?

kennytm replied:
well it's being removed rust-lang/rust#130780

```rust
#[cfg(all(target_os = "linux", target_arch = "arm"))]
#[cfg(target_os("linux") && target_arch("arm"))]
#[cfg(target(os = "linux", arch = "arm"))]
```

Urgau replied:
Last comments from T-lang members tend to the opposite, so probably going to stay.

tmandry commented:
Things I'm quite sure I like:

* `&&` instead of `all(...)`. It's too difficult to find a visual distinction between `all` and `any` and I often miss it.
* `feature("string")` instead of `feature = "string"`. This was already confusing to begin with.

Things I'm less sure of but still open to:

* `||` instead of `any(...)`. We should probably allow it for consistency, but I do find myself agreeing with @RalfJung that it ends up looking nicer to use `any()`. The main thing for me is having a clear visual distinction between the two, as I said above, and we technically only need one operator to accomplish that.
* `!`. I think this will be much improved when combined with `feature("string")`. I like that it's "loud" visually, but most times it has to be used like `!()` which never feels that great.
* Incorporating pattern syntax. We might need to extend it beyond `|` to include `&`. This does seem like it could be a win, but it implies to me that we would want to unify these as far as we sensibly could. I'd want to discuss the possible implications of that in more depth.

tgross35 commented:
When deciding upon a pattern here, it's probably worth thinking about the standard library's close-to-stable [`cfg_match!`](https://github.com/rust-lang/rust/issues/115585). It would be great if the same syntax was usable in both `cfg(...)` and `cfg_match!{ ... }`.

kennytm commented:
since `cfg_match!` is just [implemented as a decl-macro](https://github.com/rust-lang/rust/blob/847e3ee6b0e614937eee4e6d8f61094411eadcc0/library/core/src/macros/mod.rs#L238-L259) that expands to `#[cfg]` i don't see why a syntax is only usable in one place but not the other.

kennytm commented:
The problem adapting to `cfg_match!` is that `a && b` etc. are not valid `:meta` in a macro. This is not a problem of `cfg_match!` alone, any other decl-macro expecting `#[cfg($cfg:meta)]` will be affected

joshtriplett commented:
@rust-lang/lang, shall we accept the RFC as currently proposed, for adding `&&`, `||`, and `!` to `cfg` syntax?

As currently written, the RFC defers all other potential additions to future proposals.

@rfcbot merge

jhpratt commented:
That is my understanding of the current state as well. If accepted, I will almost certainly follow up with a function-like `feature("foo")` proposal as a standalone RFC.

Thanks for moving this forward, Josh!

tgross35 commented on text/0000-cfg-logical-ops.md:
```diff
@@ -0,0 +1,146 @@
+| `a \|\| b && c \|\| d`             | `any(a, all(b, c), d)`                             | `\|\|` has lower precedence than `&&`                                               |
+| `(a \|\| b) && (c \|\| d)`         | `all(any(a, b), any(c, d))`                        | `()` can be used for grouping                                                       |
+| `!a \|\| !b && !c`                 | `any(not(a), all(not(b), not(c)))`                 | `!` has highest precedence                                                          |
+| `feature="foo" \|\| feature="bar"` | `any(feature="foo", feature="bar")`                | `\|\|` has lower precedence than `=`                                                |
+| `feature="foo" && feature="bar"`   | `all(feature="foo", feature="bar")`                | `&&` has lower precedence than `=`                                                  |
+| `!feature="foo"`                   | _syntax error_                                     | `!` has higher precedence than `=`, which may be confusing, so we ban this syntax   |
+| `!(feature="foo")`                 | `not(feature="foo")`                               | use `()` for grouping                                                               |
```
(starting an inline discussion for brainstorming `key = "val"` syntax stemming from https://github.com/rust-lang/rfcs/pull/3796#issuecomment-2993831077)

> > FWIW I find `S(e)` to be a quite natural way to represent `e ∈ S`, but that is probably because of my background in type theory where sets are often modeled as predicates ("functions with return type `bool`", basically), and so asking "is this value in the set" is literally a function call.
> 
> Yes; I have a similar intuition for it, and agree that it's natural in that way. I'm just not sure about putting this set-membership predicate in the same syntax space as arbitrary predicates, and thereby making it a breaking change to add any new such predicates.

Could we add a new named predicate for this? `set(feature, "foo")` is terse and could work, though I don't love that the intuition here would probably indicate a 1:1 mapping. `contains` or `includes` are probably how we would spell this on a collection, but I think `contains(feature, "foo")` sounds a bit worse than plural would be `contains(features, "foo")`.

jhpratt replied:
I'll have an RFC together soon, though deciding what the syntax should be will definitely be better before it's fully written.

Random thought, which admittedly isn't very Rust-like, is `feature?("foo")`, along with `feature?("foo" | "bar")` and `feature?("foo" & "bar")`. There is some precedence in Ruby where `?` means a function that returns a boolean.

kornelski replied:
Rust typically uses `foo()` for getters and `set_foo()` for setters, and `cfg()` is a context that is querying the config rather than setting it, so to be this doesn't look like a setter. 

`has_feature()` would be even clearer.

jmjoy commented:
Simple Boolean operations are one thing, but what I’m really hoping for is the ability to compare values, something that’s quite common in C macros, like:

`#if FOO > 1000`

But in Rust, there’s currently no way to express this, which leads to some rather awkward workarounds.

Right now, the values in Rust’s `cfg` are all strings, so it seems difficult to support something like:

`#[cfg(FOO > 1000)]`

Jules-Bertholet commented:
It would be nice if this RFC also specified a new `macro_rules!` fragment specifier to match an arbitrary `cfg` predicate, for use by macros like `cfg_if`. Currently, you can use `meta`, which excludes only `true` and `false`, but with this RFC, that would become completely inadequate.

Jules-Bertholet commented:
> Sure, things like `cfg_if` and `cfg_select` need to work around some things because they match on the inner parts, but isn't that already the case?

Right now, the only thing they “need to work around” is `true`/`false`:

- `cfg_if` simply [doesn’t support them](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=d314684c216d5241bc19e8eb8e5e5660).
- `cfg_select` in the stdlib [doesn’t use `macro_rules!`](https://doc.rust-lang.org/src/core/macros/mod.rs.html#238).
- https://github.com/rust-lang/rust/pull/146281 (a PR I wrote that modifies the std `thread_local` such that it parses `#[cfg_attr(…)]`) uses [extra](https://github.com/rust-lang/rust/blob/1bfbfbfa0e41e5d7fb9a257199eecb7d828de1f1/library/std/src/sys/thread_local/os.rs#L75-L91) [match](https://github.com/rust-lang/rust/blob/1bfbfbfa0e41e5d7fb9a257199eecb7d828de1f1/library/std/src/thread/local.rs#L196-L224) [arms](https://github.com/rust-lang/rust/blob/1bfbfbfa0e41e5d7fb9a257199eecb7d828de1f1/library/std/src/thread/local.rs#L266-L284) just for `true` and `false`, which is annoying but doable. Supporting this RFC’s syntax, in contrast, would require complex TT-munching, which is *really* unappealing, especially given that the macro has far too much of that going on already.

jhpratt commented:
How common are macros that need to handle the _internal_ part of `cfg`? Ultimately that's what we're talking about here. The fact that `meta`, prior to `true`/`false`, matched the internal part perfectly was largely a historical artifact from before arbitrary tokens could be accepted. That restriction was lifted in April 2019.

Essentially, my argument is that this scenario is sufficiently uncommon that it is not worth the added language complexity for a rarely used macro fragment specifier.

Jules-Bertholet commented:
> How common are macros that need to handle the _internal_ part of cfg?

I just cited three examples.

If you feel that defining an entirely new specifier is too much complexity, one alternative would be to re-use `expr`.

Jules-Bertholet commented:
Actually, I realized that `cfg_if` doesn’t need a dedicated fragment specifier, it can just switch to `tt`: https://github.com/rust-lang/cfg-if/pull/99. That wouldn’t work for `cfg_match` or https://github.com/rust-lang/rust/pull/146281, however.

Jules-Bertholet commented:
> Increasingly I am thinking that this is probably the best solution. `expr` already accepts all the syntax in this RFC, the restriction on its use in `cfg` simply needs to be lifted.

https://github.com/rust-lang/rust/pull/146961

jhpratt commented:
Regarding precedence, would it be easier to pass the RFC if instead we required `(feature = "foo")`, erroring on the unparenthesized version? That would avoid the precedence carveout while being forward compatible with anything we want.

jhpratt committed:
Remove precedence carveout
